src/adapter/http/src/upload/upload_service.rs

```diff
@@ -0,0 +1,197 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use base64::Engine;
+use bytes::Bytes;
+use kamu::domain::{InternalError, MediaType, ResultIntoInternal};
+use opendatafabric::AccountID;
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use thiserror::Error;
+use tokio::io::AsyncRead;
+
+///////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+pub trait UploadService: Send + Sync {
+    async fn make_upload_context(
+        &self,
+        account_id: &AccountID,
+        file_name: String,
+        content_type: String,
+        content_length: usize,
+    ) -> Result<UploadContext, MakeUploadContextError>;
+
+    async fn upload_token_into_stream(
+        &self,
+        account_id: &AccountID,
+        upload_token: &str,
+    ) -> Result<(Box<dyn AsyncRead + Send + Unpin>, MediaType), UploadTokenIntoStreamError> {
+        let upload_token_payload = decode_upload_token_payload(upload_token)
```

Comment:
Should this be moved into `UploadServiceImpl`? It seems that after we added this high-level method `UploadService` trait doesn't need to provide the low-level operations like `upload_reference_size` and `upload_reference_into_stream` - those can be made private methods of implementation. Or do you see them useful?

Reply:
I guess this provides some reuse between local / S3 implementations, but mb this could be done through a `common` function. This kind of logic seems not a "trait material" to me, but will leave the decision up to you.

Reply:
Made it a global function in the same module as `UploadService`.